### PR TITLE
gnrc_ipv6_nc: introduce L2 address lookup function

### PR DESCRIPTION
**Pull request by authmillenon**
Removes some code duplication.

### CODE REVIEW

File: sys/include/net/gnrc/ipv6/nc.h [hunk 1]
```diff
@@ -293,6 +294,29 @@ static inline bool gnrc_ipv6_nc_is_reachable(const gnrc_ipv6_nc_t *entry)
...
+ * @return  PID to the interface where the neighbor is.
+ */
+static inline kernel_pid_t gnrc_ipv6_nc_get_l2_addr(uint8_t *l2_addr, uint8_t *l2_addr_len,
+                                                    const gnrc_ipv6_nc_t *entry)
+{
+    assert((l2_addr != NULL) && (l2_addr_len != NULL));
+    if ((entry == NULL) || !gnrc_ipv6_nc_is_reachable(entry)) {
+        return KERNEL_PID_UNDEF;
+    }
+    *l2addr_len = entry->l2_addr_len;
```
Comment (OlegHahm) [inline sys/include/net/gnrc/ipv6/nc.h]: `s/l2addr_len/l2_addr_len/`

File: sys/include/net/gnrc/ipv6/nc.h [hunk 2]
```diff
@@ -293,6 +294,29 @@ static inline bool gnrc_ipv6_nc_is_reachable(const gnrc_ipv6_nc_t *entry)
...
+    memcpy(l2addr, entry->l2_addr, entry->l2_addr_len);
```
Comment (OlegHahm) [inline sys/include/net/gnrc/ipv6/nc.h]: `s/l2addr/l2_addr/`

File: sys/include/net/gnrc/ipv6/nc.h [hunk 3]
```diff
@@ -21,6 +21,7 @@
 #ifndef GNRC_IPV6_NC_H_
 #define GNRC_IPV6_NC_H_
 
+#include <assert.h>
```
Comment (OlegHahm) [inline sys/include/net/gnrc/ipv6/nc.h]: `<string.h>` is required for memcpy, too.

**Comment by authmillenon** [comment]
Addressed comments.

**Comment by OlegHahm** [comment]
My tests yesterday (with doing the required changes myself) worked fine, but I would feel more comfortable if you, @authmillenon or anyone else (@cgundogan, @BytesGalore?) could confirm positive testing results, too.

**Comment by authmillenon** [comment]
Provided some unittests.

File: sys/include/net/gnrc/ipv6/nc.h (hunk 3 above)
Comment (authmillenon) [inline sys/include/net/gnrc/ipv6/nc.h]: (see below)

**Comment by authmillenon** [comment]
And updated doc, to reflect tests better.

**Comment by OlegHahm** [comment]
Did anyone test in the meantime?

**Comment by authmillenon** [comment]
I added the unittests, they work.

**Comment by authmillenon** [comment]
Or did you mean runs of the network stack?

**Comment by OlegHahm** [comment]
The latter.

**Comment by authmillenon** [comment]
No I did not test this yet. :/

**Comment by authmillenon** [comment]
Rebased and tested with

- [x] gnrc_networking on native
- [ ] gnrc_networking without neighbor discovery on native
- [ ] gnrc_networking on iotlab-m3

**Comment by authmillenon** [comment]
Patch for the "gnrc_networking without neighbor discovery on native" was

```diff
diff --git a/examples/gnrc_networking/Makefile b/examples/gnrc_networking/Makefile
index b73bb4d..ceef7f4 100644
--- a/examples/gnrc_networking/Makefile
+++ b/examples/gnrc_networking/Makefile
@@ -17,10 +17,8 @@ BOARD_INSUFFICIENT_MEMORY := airfy-beacon chronos msb-430 msb-430h nrf51dongle \
 USEMODULE += gnrc_netif_default
 USEMODULE += auto_init_gnrc_netif
 # Specify the mandatory networking modules for IPv6 and UDP
-USEMODULE += gnrc_ipv6_router_default
+USEMODULE += gnrc_ipv6_router
 USEMODULE += gnrc_udp
-# Add a routing protocol
-USEMODULE += gnrc_rpl
 # This application dumps received packets to STDIO using the pktdump module
 USEMODULE += gnrc_pktdump
 # Additional networking modules that can be dropped if not needed
```

**Comment by authmillenon** [comment]
Arghs, somehow all the changes disappeared...

**Comment by authmillenon** [comment]
Tests are working regardless :-)

**Comment by authmillenon** [comment]
(with the changes reverted that is ;-))

File: sys/include/net/gnrc/ipv6/nc.h [hunk 4]
```diff
@@ -293,6 +295,30 @@ static inline bool gnrc_ipv6_nc_is_reachable(const gnrc_ipv6_nc_t *entry)
...
+ * @pre (l2_addr != NULL) && (l2_addr_len != NULL)
+ *
+ * @param[out] l2_addr      The link layer address of @p entry. Must not be NULL.
+ * @param[out] l2_addr_len  Length of @p l2_addr. Must not be NULL.
+ * @param[in] entry         A neighbor cache entry
+ *
+ * @return  PID to the interface where the neighbor is.
+ * @return  KERNEL_PID_UNDEF, if @p entry == NULL or the neighbor is not reachable.
+ */
+static inline kernel_pid_t gnrc_ipv6_nc_get_l2_addr(uint8_t *l2_addr, uint8_t *l2_addr_len,
```
Comment (cgundogan) [inline sys/include/net/gnrc/ipv6/nc.h]: I would vote for not `inlining` this function. I can imagine that `gnrc_ipv6_nc_get_l2_addr()` will be called from different places  (nd, routing protocols, ...)

**Comment by authmillenon** [comment]
Addressed comments and squashed immediately.

**Comment by cgundogan** [comment]
ACK and GO

**Pull request merged**